### PR TITLE
New controller for flexibility trading with Levl Energy

### PR DESCRIPTION
**Pull request by DennisLevl**
# Adds new controller for flexibility trading with Levl Energy
## General Information
This pull request introduces a new controller that enables OpenEMS and Fenecon customers to trade their flexibility using the Levl Energy platform. This functionality is currently in the **beta** stage and will be refined based on feedback and additional testing.

## How it works
This controller receives charging/discharging instructions from Levl Energy and executes these taking into account local optimization and various constraints (such as state of charge, grid limits, etc.).

## Testing
The business logic has been covered by unit tests. The controller has also been tested using multiple test scenarios within the OpenEMS test framework to ensure functionality and reliability.

Please review the changes and provide feedback. Once approved, this will be ready for merge. Thank you in advance!

### CODE REVIEW

File: io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/ControllerEssBalancingImpl.java [hunk 1]
```diff
@@ -0,0 +1,583 @@
...
+		var levlSocWs = this.getLevlSoc().getOrError();
+		var remainingLevlEnergyWs = this.getRemainingLevlEnergy().getOrError();
+		var efficiency = this.getEfficiency().getOrError();
+		var socLowerBoundLevlPercent = this.getSocLowerBoundLevl().getOrError();
+		var socUpperBoundLevlPercent = this.getSocUpperBoundLevl().getOrError();
+		var buyFromGridLimit = this.getBuyFromGridLimit().getOrError();
+		var sellToGridLimit = this.getSellToGridLimit().getOrError();
+		var influenceSellToGrid = this.getInfluenceSellToGrid().getOrError();
+
+		var essCapacityWs = essCapacity * 3600L;
```
[the remark this answers is not in the document]
Comment (DennisLevl) [reply io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/ControllerEssBalancingImpl.java]: Good point, done.

File: io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/ControllerEssBalancingImpl.java [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+
+		// primary use case (puc) calculation
+		var pucSocWs = this.calculatePucSoc(physicalSocWs, levlSocWs, essCapacityWs);
+		var pucBatteryPower = this.calculatePucBatteryPower(gridPower, essPower, pucSocWs, essCapacityWs, minEssPower,
+				maxEssPower, efficiency, cycleTimeS);
+		var pucGridPower = gridPower + essPower - pucBatteryPower;
+		var nextPucSocWs = pucSocWs - Math.round(Efficiency.apply(pucBatteryPower, efficiency) * cycleTimeS);
+
+		// levl calculation
+		var levlBatteryPower = 0;
```
[the remark this answers is not in the document]
Comment (DennisLevl) [reply io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/ControllerEssBalancingImpl.java:188]: In our implementation, we followed the OpenEMS view for signs, in which OpenEMS is located in the center of the system. If something enters the system (e.g. battery is discharged), this is positive. If it leaves the system (e.g. battery is being charged), it is negative.

Writing out Puc would be an option, but we have decided to use the abbreviation for reasons of readability, variable length and the due to the number of uses and explain it accordingly at the beginning. 
However, I am flexible about this :)

File: io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/common/Efficiency.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package io.openems.edge.levl.controller.common;
+
+public class Efficiency {
```
[the remark this answers is not in the document]
Comment (DennisLevl) [reply io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/common/Efficiency.java:3]: Good point. I also corrected the method description.

File: io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
+package io.openems.edge.levl.controller;
+
+import com.google.gson.JsonObject;
```
[the remark this answers is not in the document]
Comment (DennisLevl) [reply io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java:3]: Now I use JsonUtils within the parseFields method. However, the OpenEMS'es own JsonrpcRequest.getParams() returns an gson JsonObject.
If there's a OpenEMS way to avoid this, I would appreciate a short hint :)

File: io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/ControllerEssBalancingImpl.java [hunk 5]
```diff
@@ -0,0 +1,583 @@
...
+	 * @param efficiency    the efficiency of the system [%]
+	 * @param cycleTimeS    the configured openems cycle time [seconds]
+	 * @return the restricted pucPower [W]
+	 */
+	protected int applyPucSocBounds(int pucPower, long pucSocWs, long essCapacityWs, double efficiency,
+			double cycleTimeS) {
+		var dischargeEnergyLowerBoundWs = pucSocWs - essCapacityWs;
+		var dischargeEnergyUpperBoundWs = pucSocWs;
+
+		var powerLowerBound = Efficiency.unapply(Math.round(dischargeEnergyLowerBoundWs / cycleTimeS), efficiency);
```
[the remark this answers is not in the document]
Comment (DennisLevl) [reply io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/ControllerEssBalancingImpl.java]: Done :)

File: io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/ControllerEssBalancingImpl.java [hunk 6]
```diff
@@ -0,0 +1,583 @@
...
+	 * @param socLowerBoundLevlPercent the lower levl soc limit [%]
+	 * @param socUpperBoundLevlPercent the upper levl soc limit [%]
+	 * @param essCapacityWs            the ess capacity [Ws]
+	 * @param influenceSellToGrid      whether it's allowed to influence sell to
+	 *                                 grid
+	 * @param efficiency               the efficiency of the system [%]
+	 * @param cycleTimeS               the configured openems cycle time [seconds]
+	 * @return the levl battery power [W]
+	 */
+	protected int calculateLevlBatteryPower(long remainingLevlEnergyWs, int pucBatteryPower, int minEssPower,
```
[the remark this answers is not in the document]
Comment (DennisLevl) [reply io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/ControllerEssBalancingImpl.java:322]: I moved the return operation (`max(min(power, upperBound), lowerBound)`) into a private method called `applyBound()`. 
In my opinion, the other lines do not lend themselves to being consolidated into a method, considering readability. 
But if you want to talk about a specific point, please let me know.

Thank you for your review!

**Comment by parapluplu** [comment]
Could you please ellaborate what the following means:
> The controller has also been tested using multiple test scenarios within the OpenEMS test framework to ensure functionality and reliability.

Which test scenarios did you execute and which OpenEMS test framework do you refer to?

**Comment by DennisLevl** [comment]
> Could you please ellaborate what the following means:
> 
> > The controller has also been tested using multiple test scenarios within the OpenEMS test framework to ensure functionality and reliability.
> 
> Which test scenarios did you execute and which OpenEMS test framework do you refer to?
> 
> If you refer to the tests in `BalancingImplTest` then my question is just a misunterstanding. However, since you mentioned unit tests along with other types of tests, but I only see unit tests here, I wanted to ask this question.

Thanks for the question. Of course by "OpenEMS test framework" I meant the tests in BalancingImplTest using OpenEMS' own test framework "ControllerTest". It's described as a "generic test framework for OpenEMS". Since it tests the whole controller and multiple cycles, I mentioned it separately to differentiate between the single tested methods within ControllerEssBalancingImplTest. So yes, my description might be a bit misunderstood :-)

We also tested the controller within our own development environment using several simulated OpenEMS edge instances.

File: io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+		if (this.efficiencyPercent < 0) {
+			throw OpenemsError.JSONRPC_INVALID_MESSAGE.exception("efficiencyPercent must be > 0");
+		}
+		if (this.efficiencyPercent > 100) {
+			throw OpenemsError.JSONRPC_INVALID_MESSAGE.exception("efficiencyPercent must be <= 100");
+		}
+	}
+
+	//Just for testing
+	public LevlControlRequest(int sellToGridLimitW, int buyFromGridLimitW, String levlRequestId, String timestamp,
```
Comment (parapluplu) [inline io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java:51]: package level visibility should be good enough then, isn't it?

File: io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+import io.openems.common.exceptions.OpenemsError;
+import io.openems.common.exceptions.OpenemsError.OpenemsNamedException;
+import io.openems.common.jsonrpc.base.JsonrpcRequest;
+import io.openems.common.utils.JsonUtils;
+
+import java.time.Clock;
+import java.time.LocalDateTime;
+import java.util.Objects;
+
+public class LevlControlRequest {
```
Comment (parapluplu) [inline io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java:13]: Would be good to put a version on this request, so it's easier to change the API version eventually

Comment (parapluplu) [inline io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java:13]: What's the motivation to perform parts of the controlling via an external API and not within OpenEMS?

Comment (parapluplu) [reply io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java:13]: Yes, it's about virtual power plants. I can imagine this being implemented within OpenEMS, with a virtual power plant coordinator running OpenEMS and multiple batteries connected to the power plant through OpenEMS. That’s the reason for my question.

This question isn’t intended to block or hinder the merge request; it’s simply out of curiosity.

Comment (DennisLevl) [reply io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java:13]: You're right. It's definitely something we'll add at a later date. For the beta it will be fine.

File: io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java (hunk 7 above)
Comment (parapluplu) [reply io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java]: `protected` is more than package level visibility, see https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html

But `protected` is good as well :)

File: io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java (hunk 8 above)
Comment (DennisLevl) [reply io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java:13]: Hey guys, sorry for the late response. Levl enables battery owners to participate in the electricity market, making more efficient use of the batteries and reducing amortisation costs. At the same time, the electricity grid is relieved as electricity is bought when it is cheap (a lot of renewables energy in the grid) and sold when demand is high (coal and gas is needed).

Unfortunately, this is not quite so simple from a regulatory perspective, as battery storage systems behind-the-meter, especially with a primary use case such as optimising self-consumption, requires a distinction to be made between self-consumption and temporarily stored energy. This is the only way the energy supplier can correctly bill the quantities purchased. For this reason, we act as an aggregator with Levl.

In the end, this controller is comparable to various other controllers that make it possible to use an external service on the hardware and/or software side to optimise your own system.

To come back to your idea @parapluplu: If regulation were left out of the equation, that might be possible. But you would still need someone to run and manage this coordinator centrally. However, Levl's long-term goal is to enable all battery owners, regardless of the EMS used, to become part of a virtual power plant with their system. 
At the moment we're working with a lot of battery owners who are using OpenEMS as well as working directly with Fenecon. For this reason we're looking forward to becoming part of the OpenEMS community :-)

File: io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/ControllerEssBalancingImpl.java [hunk 9]
```diff
@@ -0,0 +1,593 @@
...
+import io.openems.edge.meter.api.ElectricityMeter;
+import io.openems.edge.levl.controller.common.Efficiency;
+
+import static java.lang.Math.round;
+import static java.lang.Math.min;
+import static java.lang.Math.max;
+
+@Designate(ocd = Config.class, factory = true)
+@Component(name = "Controller.Levl.Symmetric.Balancing", immediate = true, configurationPolicy = ConfigurationPolicy.REQUIRE)
+@EventTopics({ EdgeEventConstants.TOPIC_CYCLE_BEFORE_PROCESS_IMAGE, EdgeEventConstants.TOPIC_CYCLE_AFTER_WRITE, })
```
Comment (DerStoecki) [inline io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/ControllerEssBalancingImpl.java:47]: CodeFormatting, See [FeneconBatteryExample](https://github.com/OpenEMS/openems/blob/d4d0df49c36a325941706d63dae50026842944b3/io.openems.edge.battery.fenecon.home/src/io/openems/edge/battery/fenecon/home/BatteryFeneconHomeImpl.java#L75-L84)

**Review by DerStoecki — COMMENTED** [review]
Hey,
Just had a quick glance.
Here are my 2 cents, listed down below.


Also: Please consider to rename the package to something like
"io.openems.edge.controller.ess.balancing.levl"
or something like that, 
instead of io.openems.levl.controller
Helps to identify and categorize the controller

File: io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/ControllerEssBalancingImpl.java (hunk 9 above)
Comment (parapluplu) [reply io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/ControllerEssBalancingImpl.java:47]: maybe a stupid question, but if checkstyle does not complain code formatting should be ok, shouldn't it?

**Review by parapluplu — APPROVED** [review]

File: io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java (hunk 8 above)
Comment (parapluplu) [reply io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/LevlControlRequest.java:13]: I'm okay with this approach, but it eases it if it's built in from the beginning

File: io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/ControllerEssBalancingImpl.java (hunk 9 above)
Comment (clehne) [reply io.openems.edge.levl.controller/src/io/openems/edge/levl/controller/ControllerEssBalancingImpl.java:47]: @parapluplu You are right. If checkstyle does not complain, the code formatting should be ok. 

Regardless of this, using the code formatting like in [FeneconBatteryExample](https://github.com/OpenEMS/openems/blob/d4d0df49c36a325941706d63dae50026842944b3/io.openems.edge.battery.fenecon.home/src/io/openems/edge/battery/fenecon/home/BatteryFeneconHomeImpl.java#L75-L84) has some advantages. For example, for me it took 2-5s to understand the configuration of the fenecon battery. In your code it took me 15-20s. This is, because it took me some time to mentally divide the lines into several lines. This is how may brain is trained (because of other OpenEMS code). Other developers may have the same problem. 
As we all want to work efficiently on the same code base it would be nice if we format the code more or less identical. 
And in the end this issue should be reflected by an updated set of checkstyle rules..

**Comment by DennisLevl** [comment]
> Hey, Just had a quick glance. Here are my 2 cents, listed down below.
> 
> Also: Please consider to rename the package to something like "io.openems.edge.controller.ess.balancing.levl" or something like that, instead of io.openems.levl.controller Helps to identify and categorize the controller

Hey @DerStoecki, thanks for your 2 cents and advice. I have implemented your notes into the controller. Especially the TimeLeapClock was new to me, so thank you very much. I think we will use this in a later version for some further testing.

On the package renaming: We used io.openems.edge.levl.controller as the package name because we were inspired by other controllers like io.openems.edge.kostal.piko. 
io.openems.edge.levl because it's a new custom controller for the edge. 
To me, "io.openems.edge.controller.ess.balancing.levl" would suggest that it's a core module of openems, and probably a part or sub-implementation of the "io.openems.edge.controller.ess.balancing" package. 
However, if you think this would be a better name for identity and categorisation reasons, I'm free to change it.

**Comment by sfeilmeier** [comment]
Hi @DennisLevl, can you please allow Commits for me. This would allow me to effectively work on the PR. Thanks!

See for details: 
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by DennisLevl** [comment]
> Hi @DennisLevl, can you please allow Commits for me. This would allow me to effectively work on the PR. Thanks!
> 
> See for details: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

Unfortunately, this does not work for forks owned by an org (in this case, levlenergy). However, I have added you to the fork as a member.

**Review by sfeilmeier — APPROVED** [review]

**Comment by sfeilmeier** [comment]
I did a detailed review and code cleanups/stylings. I had to do some (possibly) breaking changes. Unit-Tests are still working, but I was not able to do a real test. Please @friends from Levl, test again and suggest changes if required.
